### ISSUE TITLE
CRFLayer.cpp:69] Check failed: starts[numSequences] == batchSize

### ISSUE THREAD

**Issue by yjj19920505** [PaddlePaddle/Paddle#6295]
用crf做序列标注的时候遇到了这个问题。
报的错误如下：
```
I1205 17:55:54.326232  2874 GradientMachine.cpp:85] Initing parameters..
I1205 17:56:01.181100  2874 GradientMachine.cpp:92] Init parameters done.
Pass 0, Batch 0, Cost 12.474894, {'__sum_evaluator_0__': 0.8558558821678162}
F1205 17:56:12.869208  2874 CRFLayer.cpp:69] Check failed: starts[numSequences] == batchSize (797 vs. 798)
*** Check failure stack trace: ***
    @     0x7f72874917ed  google::LogMessage::Fail()
    @     0x7f7287493b38  google::LogMessage::SendToLog()
    @     0x7f72874912db  google::LogMessage::Flush()
    @     0x7f7287494a0e  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f72870f8362  paddle::CRFLayer::forward()
    @     0x7f7287134049  paddle::NeuralNetwork::forward()
    @     0x7f72874612d0  GradientMachine::forwardBackward()
    @     0x7f7286f65394  _wrap_GradientMachine_forwardBackward
    @           0x4cb45e  PyEval_EvalFrameEx
    @           0x4c2765  PyEval_EvalCodeEx
    @           0x4ca8d1  PyEval_EvalFrameEx
    @           0x4c2765  PyEval_EvalCodeEx
    @           0x4ca099  PyEval_EvalFrameEx
    @           0x4c2765  PyEval_EvalCodeEx
    @           0x4ca099  PyEval_EvalFrameEx
    @           0x4c2765  PyEval_EvalCodeEx
    @           0x4ca8d1  PyEval_EvalFrameEx
    @           0x4c2765  PyEval_EvalCodeEx
    @           0x4c2509  PyEval_EvalCode
    @           0x4f1def  (unknown)
    @           0x4ec652  PyRun_FileExFlags
    @           0x4eae31  PyRun_SimpleFileExFlags
    @           0x49e14a  Py_Main
    @     0x7f72c90de830  __libc_start_main
    @           0x49d9d9  _start
    @              (nil)  (unknown)
Aborted (core dumped)
```

**Comment by yjj19920505** [comment]
网络很简单，特征为word 经过emb和fc之后输入给crf，如下所示：

`
import train_data
import math
import paddle.v2 as paddle
import paddle.v2.evaluator as evaluator
import numpy as np


word_dict_len, pos_dict_len, ner_dict_len, dep_dict_len, label_dict_len = train_data.get_dict_len()
print word_dict_len, label_dict_len

hidden_dim = 512
depth = 4
default_std = 1 / math.sqrt(hidden_dim) / 3.0
mix_hidden_lr = 1e-3

def lstm():
std_0 = paddle.attr.Param(initial_std=0.)
    std_default = paddle.attr.Param(initial_std=default_std)
    word = paddle.layer.data(
	name = "word",
	type = paddle.data_type.integer_value_sequence(word_dict_len),
	)

    word_emb = paddle.layer.embedding(input = word, size = 200)
    word_fc = paddle.layer.fc(input = word, size = 200)
    
    pos = paddle.layer.data(
	name = "pos",
	type = paddle.data_type.integer_value_sequence(pos_dict_len),
	)
    pos_emb = paddle.layer.embedding(input = pos, size = 32)
    pos_fc = paddle.layer.fc(input = pos_emb, size = 32)
    
    ner = paddle.layer.data(
	name = "ner",
	type = paddle.data_type.integer_value_sequence(ner_dict_len),
	)
    ner_emb = paddle.layer.embedding(input = ner, size = 16)
    ner_fc = paddle.layer.fc(input = ner_emb, size = 16)

    dep = paddle.layer.data(
	name = "dep",
	type = paddle.data_type.integer_value_sequence(dep_dict_len),
	)
    dep_emb = paddle.layer.embedding(input = dep, size = 32)
    dep_fc = paddle.layer.fc(input = dep_emb, size = 32)
 
    feature_out = paddle.layer.fc(
	#input = [pos_fc, ner_fc, dep_fc],
	input = word_fc,
	size = label_dict_len
	)
    return feature_out
def main():
    paddle.init(use_gpu = False, trainer_count = 1)
    feature_out = lstm()
    target = paddle.layer.data(name = "target", type = paddle.data_type.integer_value_sequence(label_dict_len))
    crf_cost = paddle.layer.crf(
        size=label_dict_len,
        input=feature_out,
        label=target,
        param_attr=paddle.attr.Param(
            name='crfw', initial_std=default_std, learning_rate=mix_hidden_lr))

    crf_dec = paddle.layer.crf_decoding(
        size=label_dict_len,
        input=feature_out,
        label=target,
        param_attr=paddle.attr.Param(name='crfw'))
    evaluator.sum(input=crf_dec)

    parameters = paddle.parameters.create(crf_cost)
    optimizer = paddle.optimizer.Momentum(
        momentum=0,
        learning_rate=2e-2,
        regularization=paddle.optimizer.L2Regularization(rate=8e-4),
        model_average=paddle.optimizer.ModelAverage(
            average_window=0.5, max_average_window=10000), )

    trainer = paddle.trainer.SGD(
        cost=crf_cost,
        parameters=parameters,
        update_equation=optimizer,
        extra_layers=crf_dec)
    feeding = {
	"word" : 0,
	"pos" : 1,
	"ner" : 2,
	"dep" :3,
	"target" :4
    }
    def event_handler(event):
        if isinstance(event, paddle.event.EndIteration):
            if event.batch_id % 100 == 0:
                print "Pass %d, Batch %d, Cost %f, %s" % (
                    event.pass_id, event.batch_id, event.cost, event.metrics)
            if event.batch_id % 1000 == 0:
            	with open('params_pass_%d_%d.tar' % (event.pass_id, event.batch_id), 'w') as f:
                	trainer.save_parameter_to_tar(f)

            	#result = trainer.test(reader=paddle.batch(my_data.train(), batch_size = 10), feeding = feeding)
            	#print "\nTest with Pass %d, Batch %d, %s" % (event.pass_id, event.batch_id, result.metrics)

    trainer.train(
        reader=paddle.batch(paddle.reader.shuffle(train_data.train(), buf_size = 1024), batch_size = 100),
        event_handler=event_handler,
        num_passes=10,
        feeding=feeding)


if __name__ == '__main__':
    main()
`

**Comment by QiJune** [comment]
看训练日志，已经开始一个batch的训练了。可以检查一下输入数据的格式，是否统一。

**Comment by lcy-seso** [comment]
恰好看到这个issue，上面的配置里面有一个bug，
```python
feature_out = paddle.layer.fc(
	#input = [pos_fc, ner_fc, dep_fc],
	input = word_fc,
	size = label_dict_len
)
return feature_out
```
paddle V2 fc 默认激活是 `tanh`，这个feature out 计算的是状态特征，请务必显示地指定使用线性激活。

**Comment by yjj19920505** [comment]
@QiJune 
输入的数据都是统一处理的，如果使用dep_fc,pos_fc,ner_fc作为feature_out的输入都是没问题的，我也检查了word的输入数据，也没有出现空的数据，每一个样本的word、dep、pos、ner特征的列表长度都是一样的，唯一不同之处在于word的词典数达到了10W+，而dep、pos、ner的词典数最大也就几百。

**Comment by yjj19920505** [comment]
@lcy-seso 如果feature_out采用默认的tanh激活，输入为dep、pos、ner的时候是可以训练下去的，不会出错。现在修改为线性激活后，采用word作为输入依然会报相同的错误。

**Comment by lcy-seso** [comment]
这个错误是数据的错误。线性激活不会改变数据的错误。但是如果使用tanh，这个违背了CRF的原理，会得到很糟糕的学习效果。

**Comment by yjj19920505** [comment]
@lcy-seso  好的，谢谢您的指导，因为是第一次使用crf，对它的原理不是很熟悉。

**Comment by lcy-seso** [comment]
按照这个地方，把激活显示地设置成线性的即可 https://github.com/PaddlePaddle/models/blob/develop/sequence_tagging_for_ner/network_conf.py#L86

**Comment by yjj19920505** [comment]
@lcy-seso ，我的网络是参照paddle教程的语义角色标注进行的，在这里的feature_out是采用的默认激活函数，请问为什么这里可以采用tanh激活函数呢？https://github.com/PaddlePaddle/book/blob/develop/07.label_semantic_roles/train.py

**Comment by lcy-seso** [comment]
我从代码角度解释一下上面的报错，你试试有没有线索。现在挂掉的这个检查在检查：
```cpp
CHECK_EQ(starts[numSequences], batchSize);
```
1.  `starts[numSequences]` label 里面有多少token（序列中的一个词是一个token）；
2.  batchSize 是 `feature_out` 也就是CRF的输入这个矩阵的高度；

以上两者不相等。

**Comment by lcy-seso** [comment]
> 请问为什么这里可以采用tanh激活函数呢？

这是一个bug，不应该用tanh，学习效果会很差，需要改过来。

我推测是由于意外，`fc` 层的激活，习惯上大家会默认采用线性激活，而Paddle 的 v2 默认是tanh。

**Comment by yjj19920505** [comment]
@lcy-seso 好嘞，非常感谢您的悉心指教啦O(∩_∩)O~~

**Comment by lcy-seso** [comment]
我检查了一下 book 里面的 SRL 的最新版本
 https://github.com/PaddlePaddle/book/blob/develop/07.label_semantic_roles/train.py#L103 里面的 `feature_out`层，这个层使用 `paddle.layer.mixed` 计算得到。而 `paddle.layer.mixed` 的默认激活是线性，是没有问题的。
